library/std/src/io/impls.rs

```diff
@@ -357,18 +357,26 @@ impl Write for &mut [u8] {
 
 /// Write is implemented for `Vec<u8>` by appending to the vector.
 /// The vector will grow as needed.
+///
+/// # Panics
+///
+/// In case of allocation error or capacity overflow, write operations will panic.
+/// The panicking behavior is not guaranteed. In the future, it may become
+/// a regular `io::Error`.
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<A: Allocator> Write for Vec<u8, A> {
     #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        fallible_reserve(self, buf.len());
         self.extend_from_slice(buf);
         Ok(buf.len())
     }
 
     #[inline]
     fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let len = bufs.iter().map(|b| b.len()).sum();
-        self.reserve(len);
+        let len =
+            bufs.iter().try_fold(0usize, |len, b| len.checked_add(b.len())).expect("out of memory");
```

Comment:
I don't understand why is it out of memory when checked_add fails? Does the iterator perhaps allocate?

Reply:
Strictly speaking it's an equivalent of `AllocError::CapacityOverflow`, but I think the distinction is not important, and exists in Rust only because of history of its particular implementation.

1. If the preallocation check didn't exist, the same inputs would have resulted in OOM. When the sum of necessary allocations can't fit in `usize`, we know for sure it would OOM at some point, so it can be interpreted as an "as-if" equivalence.

2. This may eventually become `io::ErrorKind::OutOfMemory`. There's no `ErrorKind` for `CapacityOverflow`, and probably shouldn't be, so there's no point introducing this distinction temporarily.